src/Model/index.js

```diff
@@ -118,6 +120,28 @@ export default class Model {
     return this
   }
 
+  prepareMarkAsDeleted(): this {
+    invariant(this._isCommitted, `Cannot mark an uncomitted record as deleted`)
+
+    this._isEditing = true
+    this._raw._status = 'deleted'
+    this._hasPendingDelete = 'mark'
+    this._isEditing = false
+
+    return this
+  }
+
+  prepareDestroyPermanently(): this {
+    invariant(this._isCommitted, `Cannot mark an uncomitted record as deleted`)
+
+    this._isEditing = true
+    this._raw._status = 'deleted'
+    this._hasPendingDelete = 'destroy'
+    this._isEditing = false
+
+    return this
+  }
+
   // Marks this record as deleted (will be permanently deleted after sync)
   // Note: Use this only with Sync
   async markAsDeleted(): Promise<void> {
```

Comment:
I think most of markAsDeleted/destroyPermanently methods should be replaced with a call to database.batch(prepareXxxx) — similar to how model.update mostly just does database.batch(model.prepareUpdate). This is to avoid duplicate code and two code paths doing essentially the same thing. There's also some support code in Collection we could then delete

Reply:
Roger that. I will try to put in some finishing touches that accounts for this

Reply:
awesome 👍